### PR TITLE
[store] row length guards for prolly serial messages

### PR DESCRIPTION
**Pull request by max-hoffman**
Prevent writing rows that exceed the prolly serial message max size. When writing fields we use the current buffer size, the new field length, and the serial message metadata size to validate inserts.

fixes: https://github.com/dolthub/dolt/issues/7487
fixes: https://github.com/dolthub/dolt/issues/7524

companion: https://github.com/dolthub/go-mysql-server/pull/2342

### CODE REVIEW

File: go/store/val/tuple.go [hunk 1]
```diff
@@ -15,14 +15,22 @@
...
 )
 
 const (
-	MaxTupleFields            = 4096
-	MaxTupleDataSize ByteSize = math.MaxUint16
+	MaxTupleFields = 4096
+
+	// MaxTupleDataSize is the maximum KV length considering the extra
+	// flatbuffer metadata required to serialize the message. This number
+	// is only useful for checking the "last row" per-es, because every field
```
Comment (nicktobey) [inline go/store/val/tuple.go:29]: "per-es" is this a typo?

File: go/store/val/tuple.go [hunk 2]
```diff
@@ -15,14 +15,22 @@
...
+	// has offsets but field count, content hash, etc are global properties
+	// of the message.
+	// (uint16) - (2 kv offsets) - (field count) - (content hash) - (node count) - (tree level)
```
Comment (nicktobey) [inline go/store/val/tuple.go:32]: Not sure I understand why there are exactly two kv offsets in the message?

**Comment by max-hoffman** [comment]
@nicktobey I made two notable changes
-  the row length check is now a responsibility of the integrator. The memory db simulates closer to what innoDB does, and Dolt's limit will be a bit less because our refs are addresses rather than pointers and flatbuffers have extra metadata.
- I tried naming the metadata offsets so they are more self-documenting, and moved the field-specific offsets to PutField which has the ability to skip the first row's offsets

**Review by nicktobey — APPROVED** [review]

**Review by nicktobey — APPROVED** [review]

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8cc4acd</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8cc4acd</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>659999e</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>659999e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>ea92f6b</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>ea92f6b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8a6c533</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8a6c533</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>2a4f295</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>2a4f295</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>01100cd</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>01100cd</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**